# DATA WRANGLING PROJECT

## Problem statement: In this part of the project, we want to find out the following information;
**- How many people responded to the questions.** <br>

**- How many answers were given to these questions.**

**- How many people gave the same answer per question**

We'll be making use of the pandas library.
## Let's import the librairies and import our dataset.

In [1]:
import pandas as pd

In [2]:
dataset = pd.read_excel("Survey Monkey Output (My Edited).xlsx", sheet_name="Edited_Data")
dataset.head()

Respondent ID          Start Date            End Date  Email Address  \
0     5379192392 2021-01-22 12:01:17 2021-01-22 12:40:34            NaN   
1     2658722536 2021-01-22 06:56:37 2021-01-22 07:34:10            NaN   
2     4044163394 2021-01-22 06:35:18 2021-01-22 06:47:32            NaN   
3     5535865599 2021-01-21 21:29:32 2021-01-21 21:40:24            NaN   
4     3356802928 2021-01-21 17:26:39 2021-01-21 17:44:40            NaN   

   First Name  Last Name  Custom Data 1-Response  \
0         NaN        NaN                     NaN   
1         NaN        NaN                     NaN   
2         NaN        NaN                     NaN   
3         NaN        NaN                     NaN   
4         NaN        NaN                     NaN   

  Identify which division you work in.-Other (please specify)  \
0                                     Infrastructure            
1                                            Finance            
2                                     Infrastructure            
3                                     Infrastructure            
4                                    Port Operations            

  Identify which division you work in.-Response  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   

  Which of the following best describes your position level?-Response  ...  \
0                                              Staff                   ...   
1                                              Staff                   ...   
2                                    Department Lead                   ...   
3                                            Manager                   ...   
4                                            Manager                   ...   

  Question 29-Response 9 Question 29-Response 10 Question 29-Response 11  \
0                    NaN                Answer 8                Answer 8   
1                    NaN                Answer 5                     NaN   
2                    NaN                     NaN                Answer 4   
3               Answer 2                Answer 5                Answer 7   
4                    NaN                Answer 5                Answer 4   

  Question 29-Response 12 Question 29-Response 13 Question 29-Response 14  \
0                Answer 4                     NaN                     NaN   
1                     NaN                Answer 2                     NaN   
2                Answer 4                Answer 6                     NaN   
3                     NaN                Answer 6                     NaN   
4                Answer 4                Answer 7                Answer 7   

  Question 29-Response 1 Question 30-Response 2 Question 30-Response 3  \
0               Answer 5                    NaN                    NaN   
1               Answer 5                    NaN                    NaN   
2               Answer 6                    NaN               Answer 1   
3               Answer 7               Answer 7               Answer 1   
4                    NaN               Answer 7                    NaN   

  Unnamed: 99  
0         NaN  
1    Answer 1  
2         NaN  
3    Answer 6  
4    Answer 8  

[5 rows x 100 columns]

### Inspecting the dataset

In [3]:
print("The length of the dataset is: ", len(dataset))

The length of the dataset is:  198


You can see that the dataset has just 198 rows. That's because of its wide current format. 
When it's been wrangled, it'll be way more than that.

## I'll make an extra copy of the dataset

In [4]:
data= dataset.copy()
data.head()

Respondent ID          Start Date            End Date  Email Address  \
0     5379192392 2021-01-22 12:01:17 2021-01-22 12:40:34            NaN   
1     2658722536 2021-01-22 06:56:37 2021-01-22 07:34:10            NaN   
2     4044163394 2021-01-22 06:35:18 2021-01-22 06:47:32            NaN   
3     5535865599 2021-01-21 21:29:32 2021-01-21 21:40:24            NaN   
4     3356802928 2021-01-21 17:26:39 2021-01-21 17:44:40            NaN   

   First Name  Last Name  Custom Data 1-Response  \
0         NaN        NaN                     NaN   
1         NaN        NaN                     NaN   
2         NaN        NaN                     NaN   
3         NaN        NaN                     NaN   
4         NaN        NaN                     NaN   

  Identify which division you work in.-Other (please specify)  \
0                                     Infrastructure            
1                                            Finance            
2                                     Infrastructure            
3                                     Infrastructure            
4                                    Port Operations            

  Identify which division you work in.-Response  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   

  Which of the following best describes your position level?-Response  ...  \
0                                              Staff                   ...   
1                                              Staff                   ...   
2                                    Department Lead                   ...   
3                                            Manager                   ...   
4                                            Manager                   ...   

  Question 29-Response 9 Question 29-Response 10 Question 29-Response 11  \
0                    NaN                Answer 8                Answer 8   
1                    NaN                Answer 5                     NaN   
2                    NaN                     NaN                Answer 4   
3               Answer 2                Answer 5                Answer 7   
4                    NaN                Answer 5                Answer 4   

  Question 29-Response 12 Question 29-Response 13 Question 29-Response 14  \
0                Answer 4                     NaN                     NaN   
1                     NaN                Answer 2                     NaN   
2                Answer 4                Answer 6                     NaN   
3                     NaN                Answer 6                     NaN   
4                Answer 4                Answer 7                Answer 7   

  Question 29-Response 1 Question 30-Response 2 Question 30-Response 3  \
0               Answer 5                    NaN                    NaN   
1               Answer 5                    NaN                    NaN   
2               Answer 6                    NaN               Answer 1   
3               Answer 7               Answer 7               Answer 1   
4                    NaN               Answer 7                    NaN   

  Unnamed: 99  
0         NaN  
1    Answer 1  
2         NaN  
3    Answer 6  
4    Answer 8  

[5 rows x 100 columns]

We won't be needing some columns, so let's drop them.

In [5]:
columns_to_drop = ['Start Date', 'End Date', 'Email Address', 'First Name', 'Last Name', 'Custom Data 1-Response']
columns_to_drop

['Start Date',
 'End Date',
 'Email Address',
 'First Name',
 'Last Name',
 'Custom Data 1-Response']

In [6]:
data = data.drop(columns=columns_to_drop)
data.head()

Respondent ID Identify which division you work in.-Other (please specify)  \
0     5379192392                                     Infrastructure            
1     2658722536                                            Finance            
2     4044163394                                     Infrastructure            
3     5535865599                                     Infrastructure            
4     3356802928                                    Port Operations            

  Identify which division you work in.-Response  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   

  Which of the following best describes your position level?-Response  \
0                                              Staff                    
1                                              Staff                    
2                                    Department Lead                    
3                                            Manager                    
4                                            Manager                    

  Which generation are you apart of?-Response  \
0       Generation X (born between 1965-1980)   
1                                         NaN   
2       Generation X (born between 1965-1980)   
3         Millennial (born between 1981-2000)   
4       Generation X (born between 1965-1980)   

  Please select the gender in which you identify.-Response  \
0                                               Male         
1                                                NaN         
2                                               Male         
3                                         Non-Binary         
4                                             Female         

  Which duration range best aligns with your tenure at your company?-Response  \
0                                          0-2 years                            
1                                          10+ years                            
2                                          3-5 years                            
3                                         5-10 years                            
4                                          10+ years                            

  Which of the following best describes your employment type?-Response  \
0                                 Full time Employee                     
1                                 Full time Employee                     
2                                 Full time Employee                     
3                                 Full time Employee                     
4                                 Full time Employee                     

  Question 1-Response Question 2-Open-Ended Response  ...  \
0                 NaN                       Answer 6  ...   
1            Answer 4                       Answer 2  ...   
2            Answer 5                       Answer 7  ...   
3            Answer 1                       Answer 1  ...   
4                 NaN                       Answer 3  ...   

  Question 29-Response 9 Question 29-Response 10 Question 29-Response 11  \
0                    NaN                Answer 8                Answer 8   
1                    NaN                Answer 5                     NaN   
2                    NaN                     NaN                Answer 4   
3               Answer 2                Answer 5                Answer 7   
4                    NaN                Answer 5                Answer 4   

  Question 29-Response 12 Question 29-Response 13 Question 29-Response 14  \
0                Answer 4                     NaN                     NaN   
1                     NaN                Answer 2                     NaN   
2                Answer 4                Answer 6                     NaN   
3                     NaN                Ans

## Now we will to use pandas' melt method to transpose the dataset. This will make it easier to work with.
Let's create the id_vars and value_vars parameter that will be required to implement the melt method.

In [7]:
id_variables = list(data.columns)[ : 8]
value_variables = list(data.columns)[8 : ]
# value_vars

In [8]:
dataset_melted = data.melt(id_vars=id_variables, value_vars = value_variables, var_name="Question + Subquestion", value_name="Answer")
dataset_melted

Respondent ID  \
0         5379192392   
1         2658722536   
2         4044163394   
3         5535865599   
4         3356802928   
...              ...   
17023     7940065082   
17024     5157705612   
17025     9920755555   
17026     6638341389   
17027     8114622230   

      Identify which division you work in.-Other (please specify)  \
0                                         Infrastructure            
1                                                Finance            
2                                         Infrastructure            
3                                         Infrastructure            
4                                        Port Operations            
...                                                  ...            
17023                                     Infrastructure            
17024                                            Finance            
17025                                    Port Operations            
17026                                     Infrastructure            
17027                             Information Technology            

      Identify which division you work in.-Response  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   
...                                             ...   
17023                                           NaN   
17024                                           NaN   
17025                                           NaN   
17026                                           NaN   
17027                                           NaN   

      Which of the following best describes your position level?-Response  \
0                                                  Staff                    
1                                                  Staff                    
2                                        Department Lead                    
3                                                Manager                    
4                                                Manager                    
...                                                  ...                    
17023                                    Department Lead                    
17024                                              Staff                    
17025                                              Staff                    
17026                                            Manager                    
17027                                              Staff                    

      Which generation are you apart of?-Response  \
0           Generation X (born between 1965-1980)   
1                                             NaN   
2           Generation X (born between 1965-1980)   
3             Millennial (born between 1981-2000)   
4           Generation X (born between 1965-1980)   
...                                           ...   
17023        Baby Boomer (born between 1946-1964)   
17024         Millennial (born between 1981-2000)   
17025         Millennial (born between 1981-2000)   
17026         Millennial (born between 1981-2000)   
17027                        Prefer not to answer   

      Please select the gender in which you identify.-Response  \
0                                                   Male         
1                                                    NaN         
2                                                   Male         
3                                             Non-Binary         
4                                                 Female         
...                                                  ...         
17023                                               Male         
17024                                             Female         
17025                                             Female         
17026 

As we can see, the number of rows of the data is no longer 198 rows but 17000 +... after melting.

### Let's import our **questions dataset** so we can merge it to the melted data.

In [9]:
imported_questions = pd.read_excel("Survey Monkey Output (My Edited).xlsx", sheet_name="Questions")
imported_questions.head()

Raw Question Raw Subquestion       Question Subquestion  \
0  Respondent ID             NaN  Respondent ID         NaN   
1     Start Date             NaN     Start Date         NaN   
2       End Date             NaN       End Date         NaN   
3  Email Address             NaN  Email Address         NaN   
4     First Name             NaN     First Name         NaN   

  Question + Subquestion  
0          Respondent ID  
1             Start Date  
2               End Date  
3          Email Address  
4             First Name

**Let's make a copy and also drop the columns we don't need. We only need the Question and Question + Subquestion columns.**

In [10]:
questions = imported_questions.copy()
questions.drop(columns=["Raw Question", "Raw Subquestion", "Subquestion"], inplace=True)

Let's inspect the dataset and drop the missing values if there's any.

In [11]:
questions.head()

Question Question + Subquestion
0  Respondent ID          Respondent ID
1     Start Date             Start Date
2       End Date               End Date
3  Email Address          Email Address
4     First Name             First Name

In [12]:
# Dropping Missing Values
questions.dropna(inplace=True)

In [13]:
questions

Question   Question + Subquestion
0   Respondent ID            Respondent ID
1      Start Date               Start Date
2        End Date                 End Date
3   Email Address            Email Address
4      First Name               First Name
..            ...                      ...
94    Question 29  Question 29-Response 13
95    Question 29  Question 29-Response 14
96    Question 29   Question 29-Response 1
97    Question 30   Question 30-Response 2
98    Question 30   Question 30-Response 3

[99 rows x 2 columns]

## Merging the two datasets. 
The questions and the dataset we initially melted needs to be merged so we can answer the required questions.

In [14]:
merged_dataset = pd.merge(left=dataset_melted, right=questions, how="left", left_on="Question + Subquestion", right_on="Question + Subquestion")
print("Original Data", len(dataset_melted))
print("Merged Data", len(merged_dataset))
merged_dataset.head()

Original Data 17028
Merged Data 17028


Respondent ID Identify which division you work in.-Other (please specify)  \
0     5379192392                                     Infrastructure            
1     2658722536                                            Finance            
2     4044163394                                     Infrastructure            
3     5535865599                                     Infrastructure            
4     3356802928                                    Port Operations            

  Identify which division you work in.-Response  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   

  Which of the following best describes your position level?-Response  \
0                                              Staff                    
1                                              Staff                    
2                                    Department Lead                    
3                                            Manager                    
4                                            Manager                    

  Which generation are you apart of?-Response  \
0       Generation X (born between 1965-1980)   
1                                         NaN   
2       Generation X (born between 1965-1980)   
3         Millennial (born between 1981-2000)   
4       Generation X (born between 1965-1980)   

  Please select the gender in which you identify.-Response  \
0                                               Male         
1                                                NaN         
2                                               Male         
3                                         Non-Binary         
4                                             Female         

  Which duration range best aligns with your tenure at your company?-Response  \
0                                          0-2 years                            
1                                          10+ years                            
2                                          3-5 years                            
3                                         5-10 years                            
4                                          10+ years                            

  Which of the following best describes your employment type?-Response  \
0                                 Full time Employee                     
1                                 Full time Employee                     
2                                 Full time Employee                     
3                                 Full time Employee                     
4                                 Full time Employee                     

  Question + Subquestion    Answer    Question  
0    Question 1-Response       NaN  Question 1  
1    Question 1-Response  Answer 4  Question 1  
2    Question 1-Response  Answer 5  Question 1  
3    Question 1-Response  Answer 1  Question 1  
4    Question 1-Response       NaN  Question 1

### Let's perform some feature engineering.
- Check for answers that are not null
- Group questions by the number of unique respondents by their IDs
- The different repondents and their IDs.
# How many answers were given to each question?


In [18]:
respondents = merged_dataset[merged_dataset["Answer"].notna()]
respondents = respondents.groupby("Question")["Respondent ID"].nunique().reset_index()
respondents.rename(columns={"Respondent ID":"Respondents"}, inplace=True) #Renamed so it doesn't clash with the respondent ID when we want to merge the dataset.
respondents

Question  Respondents
0    Question 1          119
1   Question 10          198
2   Question 11          164
3   Question 12          114
4   Question 13          108
5   Question 14          105
6   Question 15          114
7   Question 16          117
8   Question 17          135
9   Question 18          109
10  Question 19          157
11   Question 2          118
12  Question 20          105
13  Question 21          127
14  Question 22          160
15  Question 23          120
16  Question 24          195
17  Question 25          198
18  Question 26          193
19  Question 27          153
20  Question 28          119
21  Question 29          198
22   Question 3          117
23  Question 30          160
24   Question 4          161
25   Question 5          194
26   Question 6          196
27   Question 7          162
28   Question 8          190
29   Question 9          188

As we can see from the results above, every single question was answered, which is what we actually want.
Now we want to take this dataframe of **questions and respondents** and merge with the **question** column from the merged dataset we had earlier.

In [20]:
dataset_merged2 = pd.merge(left=merged_dataset, right=respondents, how="left", left_on="Question", right_on="Question")
print("Original Data", len(merged_dataset))
print("Merged Data", len(dataset_merged2))
dataset_merged2.head()

Original Data 17028
Merged Data 17028


Respondent ID Identify which division you work in.-Other (please specify)  \
0     5379192392                                     Infrastructure            
1     2658722536                                            Finance            
2     4044163394                                     Infrastructure            
3     5535865599                                     Infrastructure            
4     3356802928                                    Port Operations            

  Identify which division you work in.-Response  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   

  Which of the following best describes your position level?-Response  \
0                                              Staff                    
1                                              Staff                    
2                                    Department Lead                    
3                                            Manager                    
4                                            Manager                    

  Which generation are you apart of?-Response  \
0       Generation X (born between 1965-1980)   
1                                         NaN   
2       Generation X (born between 1965-1980)   
3         Millennial (born between 1981-2000)   
4       Generation X (born between 1965-1980)   

  Please select the gender in which you identify.-Response  \
0                                               Male         
1                                                NaN         
2                                               Male         
3                                         Non-Binary         
4                                             Female         

  Which duration range best aligns with your tenure at your company?-Response  \
0                                          0-2 years                            
1                                          10+ years                            
2                                          3-5 years                            
3                                         5-10 years                            
4                                          10+ years                            

  Which of the following best describes your employment type?-Response  \
0                                 Full time Employee                     
1                                 Full time Employee                     
2                                 Full time Employee                     
3                                 Full time Employee                     
4                                 Full time Employee                     

  Question + Subquestion    Answer    Question  Respondents  
0    Question 1-Response       NaN  Question 1        119.0  
1    Question 1-Response  Answer 4  Question 1        119.0  
2    Question 1-Response  Answer 5  Question 1        119.0  
3    Question 1-Response  Answer 1  Question 1        119.0  
4    Question 1-Response       NaN  Question 1        119.0

The merged data and the original dataset have the same length.

 # How many people gave the same answer per question ?

In [21]:
the_same_answer = merged_dataset
the_same_answer = the_same_answer.groupby(["Question + Subquestion", "Answer"])["Respondent ID"].nunique().reset_index()
the_same_answer.rename(columns={"Respondent ID":"Same Answer"}, inplace=True)
the_same_answer

Question + Subquestion    Answer  Same Answer
0      Question 1-Response  Answer 1           14
1      Question 1-Response  Answer 2           10
2      Question 1-Response  Answer 3           13
3      Question 1-Response  Answer 4           17
4      Question 1-Response  Answer 5           22
..                     ...       ...          ...
683            Unnamed: 99  Answer 4           14
684            Unnamed: 99  Answer 5           18
685            Unnamed: 99  Answer 6           20
686            Unnamed: 99  Answer 7           15
687            Unnamed: 99  Answer 8           14

[688 rows x 3 columns]

**There are about 688 unique answers that people could have given.**
Let's carry out one more merge to put everything we've done together.

In [24]:
dataset_merged3 = pd.merge(left=dataset_merged2, right=the_same_answer, how="left", left_on=["Question + Subquestion", "Answer"], right_on=["Question + Subquestion", "Answer"])
print("Original Data: ", len(dataset_merged2))
print("Merged Data3: ", len(dataset_merged3))
dataset_merged3

Original Data:  17028
Merged Data3:  17028


Respondent ID  \
0         5379192392   
1         2658722536   
2         4044163394   
3         5535865599   
4         3356802928   
...              ...   
17023     7940065082   
17024     5157705612   
17025     9920755555   
17026     6638341389   
17027     8114622230   

      Identify which division you work in.-Other (please specify)  \
0                                         Infrastructure            
1                                                Finance            
2                                         Infrastructure            
3                                         Infrastructure            
4                                        Port Operations            
...                                                  ...            
17023                                     Infrastructure            
17024                                            Finance            
17025                                    Port Operations            
17026                                     Infrastructure            
17027                             Information Technology            

      Identify which division you work in.-Response  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   
...                                             ...   
17023                                           NaN   
17024                                           NaN   
17025                                           NaN   
17026                                           NaN   
17027                                           NaN   

      Which of the following best describes your position level?-Response  \
0                                                  Staff                    
1                                                  Staff                    
2                                        Department Lead                    
3                                                Manager                    
4                                                Manager                    
...                                                  ...                    
17023                                    Department Lead                    
17024                                              Staff                    
17025                                              Staff                    
17026                                            Manager                    
17027                                              Staff                    

      Which generation are you apart of?-Response  \
0           Generation X (born between 1965-1980)   
1                                             NaN   
2           Generation X (born between 1965-1980)   
3             Millennial (born between 1981-2000)   
4           Generation X (born between 1965-1980)   
...                                           ...   
17023        Baby Boomer (born between 1946-1964)   
17024         Millennial (born between 1981-2000)   
17025         Millennial (born between 1981-2000)   
17026         Millennial (born between 1981-2000)   
17027                        Prefer not to answer   

      Please select the gender in which you identify.-Response  \
0                                                   Male         
1                                                    NaN         
2                                                   Male         
3                                             Non-Binary         
4                                                 Female         
...                                                  ...         
17023                                               Male         
17024                                             Female         
17025                                             Female         
17026 

We can see that we stil have 17028 rows.
We can see some null values in here. Let's fill them up with zeros.

In [25]:
dataset_merged3["Same Answer"].fillna(0, inplace=True)
dataset_merged3

Respondent ID  \
0         5379192392   
1         2658722536   
2         4044163394   
3         5535865599   
4         3356802928   
...              ...   
17023     7940065082   
17024     5157705612   
17025     9920755555   
17026     6638341389   
17027     8114622230   

      Identify which division you work in.-Other (please specify)  \
0                                         Infrastructure            
1                                                Finance            
2                                         Infrastructure            
3                                         Infrastructure            
4                                        Port Operations            
...                                                  ...            
17023                                     Infrastructure            
17024                                            Finance            
17025                                    Port Operations            
17026                                     Infrastructure            
17027                             Information Technology            

      Identify which division you work in.-Response  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   
...                                             ...   
17023                                           NaN   
17024                                           NaN   
17025                                           NaN   
17026                                           NaN   
17027                                           NaN   

      Which of the following best describes your position level?-Response  \
0                                                  Staff                    
1                                                  Staff                    
2                                        Department Lead                    
3                                                Manager                    
4                                                Manager                    
...                                                  ...                    
17023                                    Department Lead                    
17024                                              Staff                    
17025                                              Staff                    
17026                                            Manager                    
17027                                              Staff                    

      Which generation are you apart of?-Response  \
0           Generation X (born between 1965-1980)   
1                                             NaN   
2           Generation X (born between 1965-1980)   
3             Millennial (born between 1981-2000)   
4           Generation X (born between 1965-1980)   
...                                           ...   
17023        Baby Boomer (born between 1946-1964)   
17024         Millennial (born between 1981-2000)   
17025         Millennial (born between 1981-2000)   
17026         Millennial (born between 1981-2000)   
17027                        Prefer not to answer   

      Please select the gender in which you identify.-Response  \
0                                                   Male         
1                                                    NaN         
2                                                   Male         
3                                             Non-Binary         
4                                                 Female         
...                                                  ...         
17023                                               Male         
17024                                             Female         
17025                                             Female         
17026 

Good Job!
Now we just need to save our data and properly rename some of these columns that are questions...
- Identify which division you work in. 
- Which of the following best describes your position level? 
- Please select the gender in which you identify  
- Which duration range best aligns with your tenure at your company
- Which of the following best describes your employment type?

In [26]:
myOutput = dataset_merged3.copy()
myOutput.rename(columns={"Identify which division you work in. - Response":"Division Primary", "Identify which division you work in. - Other (please specify)":"Division Secondary", 
    "Which of the following best describes your position level? - Response":"Position", "Which generation are you apart of? - Response":"Generation", 
    "Please select the gender in which you identify. - Response":"Gender", "Which duration range best aligns with your tenure at your company? - Response":"Tenure", 
    "Which of the following best describes your employment type? - Response":"Employment Type"}, inplace=True)
myOutput

Respondent ID  \
0         5379192392   
1         2658722536   
2         4044163394   
3         5535865599   
4         3356802928   
...              ...   
17023     7940065082   
17024     5157705612   
17025     9920755555   
17026     6638341389   
17027     8114622230   

      Identify which division you work in.-Other (please specify)  \
0                                         Infrastructure            
1                                                Finance            
2                                         Infrastructure            
3                                         Infrastructure            
4                                        Port Operations            
...                                                  ...            
17023                                     Infrastructure            
17024                                            Finance            
17025                                    Port Operations            
17026                                     Infrastructure            
17027                             Information Technology            

      Identify which division you work in.-Response  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   
...                                             ...   
17023                                           NaN   
17024                                           NaN   
17025                                           NaN   
17026                                           NaN   
17027                                           NaN   

      Which of the following best describes your position level?-Response  \
0                                                  Staff                    
1                                                  Staff                    
2                                        Department Lead                    
3                                                Manager                    
4                                                Manager                    
...                                                  ...                    
17023                                    Department Lead                    
17024                                              Staff                    
17025                                              Staff                    
17026                                            Manager                    
17027                                              Staff                    

      Which generation are you apart of?-Response  \
0           Generation X (born between 1965-1980)   
1                                             NaN   
2           Generation X (born between 1965-1980)   
3             Millennial (born between 1981-2000)   
4           Generation X (born between 1965-1980)   
...                                           ...   
17023        Baby Boomer (born between 1946-1964)   
17024         Millennial (born between 1981-2000)   
17025         Millennial (born between 1981-2000)   
17026         Millennial (born between 1981-2000)   
17027                        Prefer not to answer   

      Please select the gender in which you identify.-Response  \
0                                                   Male         
1                                                    NaN         
2                                                   Male         
3                                             Non-Binary         
4                                                 Female         
...                                                  ...         
17023                                               Male         
17024                                             Female         
17025                                             Female         
17026 

Great Job!
Let's save our final output as an excel and csv format. 
We have to add index as False so pandas doesnt create an additional index column for us.

## Summary:
We were able to successfully
- import our data in excel format using pandas
- unpivot columns (which we called melting)
- make some joins.
- drop unwanted columns and null values
- aggregate columns and do a few other things.




In [28]:
myOutput.to_excel("Final_Output.xlsx", index=False) 
myOutput.to_csv("Final_Output.csv", index=False)